src/build_system.ml

```diff
@@ -463,10 +463,14 @@ module Build_exec = struct
         | Unevaluated ->
           m.state <- Evaluating;
           let dyn_deps' = ref Pset.empty in
-          let x = exec dyn_deps' m.t x in
-          m.state <- Evaluated (x, !dyn_deps');
-          dyn_deps := Pset.union !dyn_deps !dyn_deps';
-          x
+          match exec dyn_deps' m.t x with
+          | x ->
+            m.state <- Evaluated (x, !dyn_deps');
+            dyn_deps := Pset.union !dyn_deps !dyn_deps';
+            x
+          | exception exn ->
+            m.state <- Unevaluated;
+            reraise exn
```

Comment:
Is this some bug that's being fixed? It doesn't seem related to the lib refactoring.

Reply:
Yh, it just that it popped up in a test with this PR. Basically if evaluating a memoized Build.t failed, it was staying forever in `Evaluating` state, causing a cycle error